### PR TITLE
[Driver][SYCL] device link not optimal with -save-temps enabled

### PR DESCRIPTION
**Pull request by mdtoguchi**
When using -save-temps, the temporary file names of the unbundled device
libraries did not match the template of temporary files to scrutinize
when performing the device link and adding the -only-needed flag.  Update
the matching check to work for the real temporaries and the -save-temps
variants.

### CODE REVIEW

File: clang/test/Driver/sycl-device-lib.cpp [hunk 1]
```diff
@@ -105,6 +105,8 @@
 /// test llvm-link behavior for linking device libraries
 // RUN: %clangxx -fsycl %s --sysroot=%S/Inputs/SYCL -### 2>&1 \
 // RUN:   | FileCheck %s -check-prefix=SYCL_LLVM_LINK_DEVICE_LIB
+// RUN: %clangxx -fsycl -save-temps %s --sysroot=%S/Inputs/SYCL -### 2>&1 \
+// RUN:   | FileCheck %s -check-prefix=SYCL_LLVM_LINK_DEVICE_LIB
```
Comment (AGindinson) [inline clang/test/Driver/sycl-device-lib.cpp:109]: I'd suggest a similar `sycl-device-lib-win.cpp` update for safeguarding purposes.

File: clang/lib/Driver/ToolChains/SYCL.cpp [hunk 2]
```diff
@@ -188,14 +188,14 @@ const char *SYCL::Linker::constructLLVMLinkCommand(
...
       StringRef InputFilename =
           llvm::sys::path::filename(StringRef(II.getFilename()));
-      if (!InputFilename.startswith("libsycl-") ||
+      StringRef LibSyclPrefix("libsycl-");
+      if (!InputFilename.startswith(LibSyclPrefix) ||
           !InputFilename.endswith(LibPostfix) || (InputFilename.count('-') < 2))
         return false;
-      size_t PureLibNameLen = InputFilename.find_last_of('-');
       // Skip the prefix "libsycl-"
-      StringRef PureLibName = InputFilename.substr(8, PureLibNameLen - 8);
```
Comment (AGindinson) [inline clang/lib/Driver/ToolChains/SYCL.cpp:196]: Just to confirm my understanding: had `PureLibNameLen - 8` been signifying the position of the temporary postfix?

Comment (mdtoguchi) [reply clang/lib/Driver/ToolChains/SYCL.cpp:196]: Yes, that is correct.

**Review by AGindinson — APPROVED** [review]
LGTM, thanks!

**Pull request merged**